Revamp CNI doc. (#10090)
istio/istio.io: content/en/docs/ops/diagnostic-tools/cni/index.md

---

> title: Troubleshooting CNI

Reviewer: small architecture diagram may help understand users how this works?

Author: You mean architecture for overall how istio cni works, or just log part?

Reviewer: Sounds reasonable to put on main page. I generally meant overall, but it could include the logging portion. I think most people don't actually understand CNI, especially that the daemonset doesn't really do the iptables injection itself, etc

Author: Done.

based on pod spec.

The plugin runs in the container runtime process space, so you can collect CNI plugin log from the `kubelet` log.
To ease the log collection, the CNI plugin also sends its log to the `istio-cni-node` DaemonSet.
Due to this, You can get the plugin log simply by getting `istio-cni-node` DaemonSet pod log.
The default log level for CNI plugin is at `info` level, to get more detailed log from the CNI plugin,
you can tune plugin log level via install option `values.cni.logLevel` and restart CNI DaemonSet pod.

In addition to CNI plugin log, the Istio CNI DaemonSet pod log also provides information about CNI network plugin installation,
and [race condition repairing](/docs/setup/additional-setup/cni/#race-condition-mitigation).

## Monitoring

The CNI DaemonSet [generates metrics](/docs/reference/commands/install-cni/#metrics),
which can be used to monitor CNI installation, readiness, and race condition mitigation.
Prometheus scraping annotations (`prometheus.io/port`, `prometheus.io/path`) are added to `istio-cni-node` DaemonSet pod by default.
You can collect the generated metrics via standard Prometheus configuration.

> ## Diagnose pod start-up failure

Reviewer: We should have a log of when the istio-validation container fails, and what it means.

Author: Done.

---
title: Troubleshooting CNI
description: Describes tools and techniques to diagnose issues with CNI.
weight: 90
keywords: [debug,cni]
owner: istio/wg-networking-maintainers
test: n/a
---

This page describes how to troubleshoot issues with Istio CNI.
Before reading this, you should read the [CNI installation and operation guide](/docs/setup/additional-setup/cni/).

## Log

> The Istio CNI plugin log provides information about how CNI plugin configures application pod traffic redirection
> based on pod spec.
> 
> The plugin runs in the container runtime process space, so you can collect CNI plugin log from the `kubelet` log.
> To ease the log collection, the CNI plugin also sends its log to the `istio-cni-node` DaemonSet.
> Due to this, You can get the plugin log simply by getting `istio-cni-node` DaemonSet pod log.
> The default log level for CNI plugin is at `info` level, to get more detailed log from the CNI plugin,
> you can tune plugin log level via install option `values.cni.logLevel` and restart CNI DaemonSet pod.
> 
> In addition to CNI plugin log, the Istio CNI DaemonSet pod log also provides information about CNI network plugin installation,
> and [race condition repairing](/docs/setup/additional-setup/cni/#race-condition-mitigation).

Reviewer 2: ```suggestion
The Istio CNI plugin log provides information about how the plugin configures application pod traffic redirection
based on `PodSpec`.

The plugin runs in the container runtime process space, so you can see CNI log entries in the `kubelet` log.
To make debugging easier, the CNI plugin also sends its log to the `istio-cni-node` DaemonSet.

The default log level for the CNI plugin is `info`. To get more detailed log output, you can change the level by editing the `values.cni.logLevel` installation option and restarting the CNI DaemonSet pod.

The Istio CNI DaemonSet pod log also provides information about CNI plugin installation,
and [race condition repairing](/docs/setup/additional-setup/cni/#race-condition-mitigation).
```

## Monitoring

The CNI DaemonSet [generates metrics](/docs/reference/commands/install-cni/#metrics),
which can be used to monitor CNI installation, readiness, and race condition mitigation.

> Prometheus scraping annotations (`prometheus.io/port`, `prometheus.io/path`) are added to `istio-cni-node` DaemonSet pod by default.

Reviewer 2: ```suggestion
Prometheus scraping annotations (`prometheus.io/port`, `prometheus.io/path`) are added to the `istio-cni-node` DaemonSet pod by default.
```

You can collect the generated metrics via standard Prometheus configuration.

> ## Istio CNI DaemonSet readiness
> 
> Readiness of Istio CNI DaemonSet indicates Istio CNI network plugin is properly installed and configured.
> If Istio CNI DaemonSet is unready, it suggests something is wrong with CNI plugin installation.
> Best way to check why installation fails is to look at `istio-cni-node` DaemonSet logs.
> You can also track the CNI installation readiness via metric `istio_cni_install_ready`.

Reviewer 2: ```suggestion
## DaemonSet readiness

Readiness of the CNI DaemonSet indicates that the Istio CNI plugin is properly installed and configured.
If Istio CNI DaemonSet is unready, it suggests something is wrong. Look at the  `istio-cni-node` DaemonSet logs to diagnose.
You can also track CNI installation readiness via the `istio_cni_install_ready` metric.
```

## Race condition repair

> Istio CNI DaemonSet by default has [race condition mitigation](/docs/setup/additional-setup/cni/#race-condition-mitigation) enabled,
> which will evict pod that is broken by the race condition.
> To understand which pods get evicted, you can look for log lines like the follow in CNI DaemonSet:
> 
> {{< text plain >}}
> 2021-07-21T08:32:17.362512Z     info   Deleting broken pod: service-graph00/svc00-0v1-95b5885bf-zhbzm
> {{< /text >}}
> 
> You can also track pods repaired via metric `istio_cni_repair_pods_repaired_total`.

Reviewer 2: ```suggestion
By default, the Istio CNI DaemonSet has [race condition mitigation](/docs/setup/additional-setup/cni/#race-condition-mitigation) enabled,
which will evict a pod that was started before the CNI plugin was ready.
To understand which pods were evicted, look for log lines like the following:

{{< text plain >}}
2021-07-21T08:32:17.362512Z     info   Deleting broken pod: service-graph00/svc00-0v1-95b5885bf-zhbzm
{{< /text >}}

You can also track pods repaired via the `istio_cni_repair_pods_repaired_total` metric.
```

## Diagnose pod start-up failure

> A common issue with Istio CNI is that pod fails to start up due to container network set up failure.
> Typically the failure reason is written to the pod events and is visible via pod description.
> 
> {{< text bash >}}
> $ kubectl describe pod POD_NAME -n POD_NAMESPACE
> {{< /text >}}
> 
> If pod keeps getting init error, check the init container `istio-validation` log and
> see if it complains about connection refused like the follow.
> 
> {{< text bash >}}
> $ kubectl logs POD_NAME -n POD_NAMESPACE -c istio-validation
> ...
> 2021-07-20T05:30:17.111930Z     error   Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
> 2021-07-20T05:30:18.112503Z     error   Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
> ...
> 2021-07-20T05:30:22.111676Z     error   validation timeout
> {{< /text >}}
> 
> What `istio-validation` init container does is that it sets up a local dummy server which
> listens on traffic redirection target inbound/outbound ports,
> and checks whether test traffic can be redirected to the dummy server.
> When pod traffic redirection is not set up correctly by Istio CNI,
> the `istio-validation` init container blocks pod startup to prevent traffic bypass.
> You can check `istio-cni-node` DaemonSet log and look for pod ID to see
> if there are any errors or unexpected network setup behaviors.
> 
> The other symptom of malfunctioned Istio CNI is that the application pod keeps being evicted at start-up time.
> This is typically because the CNI plugin is not properly installed, thus pod traffic redirection cannot be set up.
> CNI [race repair logic](/docs/setup/additional-setup/cni/#race-condition-mitigation) considers the pod is broken due to race condition and evicts pod continuously.
> When running into this issue, CNI DaemonSet log should have information on why CNI plugin cannot be installed properly.

Reviewer 2: ```suggestion
A common issue with the CNI plugin is that a pod fails to start due to container network set-up failure.
Typically the failure reason is written to the pod events, and is visible via pod description:

{{< text bash >}}
$ kubectl describe pod POD_NAME -n POD_NAMESPACE
{{< /text >}}

If a pod keeps getting init error, check the init container `istio-validation` log for 
"connection refused" errors like the following:

{{< text bash >}}
$ kubectl logs POD_NAME -n POD_NAMESPACE -c istio-validation
...
2021-07-20T05:30:17.111930Z     error   Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
2021-07-20T05:30:18.112503Z     error   Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
...
2021-07-20T05:30:22.111676Z     error   validation timeout
{{< /text >}}

The  `istio-validation` init container sets up a local dummy server which
listens on traffic redirection target inbound/outbound ports,
and checks whether test traffic can be redirected to the dummy server.
When pod traffic redirection is not set up correctly by the CNI plugin,
the `istio-validation` init container blocks pod startup, to prevent traffic bypass.
To see if there were any errors or unexpected network setup behaviors,
search the `istio-cni-node` for the pod ID.

Another symptom of a malfunctioned CNI plugin is that the application pod is continuously evicted at start-up time.  
This is typically because the plugin is not properly installed, thus pod traffic redirection cannot be set up.
CNI [race repair logic](/docs/setup/additional-setup/cni/#race-condition-mitigation) considers the pod is broken due to the race condition and evicts the pod continuously.
When running into this issue,  check the CNI DaemonSet log for information on why the plugin could not be properly installed.
```
